libfulltext/arxiv/__init__.py

```diff
@@ -0,0 +1,73 @@
+# copyright © 2018 the libfulltext authors (see AUTHORS.md and LICENSE)
+"""arXiv handler module"""
+
+import requests
+from lxml import etree
+
+from ..response import verify
+from ..doi import get_doi_fulltext
+
+
+def get_arxiv_fulltext(arxiv_id, save_stream, config):
+    """Get the fulltext for an arXiv ID
+
+    At the moment meant to handle one arXiv ID at a time, though the arxiv
+    API can handle multiple IDs (comma separated).
+
+    Args:
+        arxiv_id:    arxiv ID string (with prefix)
+        save_stream: function that saves a stream (arguments: stream, path)
+        config:      configuration dictionary (see config.py)
+
+    Returns:
+        What the actual getter returns (usually None)
+
+    Raises:
+        ValueError:           did not obtain exactly one query result or
+                              entry did not contain pdf link
+        NotImplementedError:  multiple arXiv IDs provided
+    """
+
+    if "," in arxiv_id:
+        raise NotImplementedError("multiple arXiv IDs not implemented")
+
+    response = requests.get("http://export.arxiv.org/api/query",
+                            params={"id_list": arxiv_id})
+    tree = etree.fromstring(response.content)
+
+    # './' = restrict search to depth 1 in the tree
+    # '{*}' = ignore the namespace (usually "http://www.w3.org/2005/Atom")
+    entries = tree.findall("./{*}entry")
+
+    if len(entries) > 1:
+        raise ValueError("Obtained more than one arXiv article")
+    elif not entries:
+        raise ValueError("Did not obtain any arXiv article")
+
+    entry = entries[0]
+
+    pdf_links = []
+    for element in entry.findall("./{*}link"):
+        if 'title' in element.keys() and element.get("title") == "pdf":
+            pdf_links.append(element.get("href"))
+
+    if not pdf_links:
+        raise ValueError("Didn't contain a pdf link")
+    elif len(pdf_links) > 1:
+        raise ValueError("Ambiguous pdf links")
+
+    response = requests.get(pdf_links[0],
+                            stream=True
+                            )
+
+    verify(response, 'application/pdf')
+
+    save_stream(response, 'arxiv.pdf')
+
+    try:
+        doi = entry.findall("./{*}doi")[0].text
```

Comment:
We shouldn't fetch the fulltext for other ids implicitly. Besides being confusing for the user it can also create loops.

Reply:
Oh yeah, it was about the likely headache in the future. :)

Reply:
Do we want to do anything else smart at this point? (have interface switch to download DOI, in the configuration, dump the DOI somewhere (txt file, so an eager user can still pick them up)?

Just got this idea, so not thought through yet, but it seems to me doing a `save_stream(<stream_that_spits_out_the_DOI>, "doi.txt")` along the third option seems nice?

Reply:
Perhaps we want to just save the metadata as well?